File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        left, right = 0, 0
        for w in weights:
            left = max(left, w)
            right += w
        while left < right:
            mid = (right-left)/2 + left
            cur, need = 0, 1
            for w in weights:
                if cur + w > mid:
                    need += 1
                    cur = 0
                cur += w
            if need > days: left = mid+1
            else: right = mid
        return left
            
        

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int left = 0, right = INT_MAX/3;
        for(int w : weights) left = max(left, w);
        
        while(left < right) {
            int mid = (right-left)/2 + left;
            int need = 1, cur = 0;
            for(int w : weights) {
                if(cur + w > mid) {
                    need++;
                    cur = 0;
                }
                cur += w;
            }
            if(need > days) left = mid+1;
            else right = mid;
        }
        
        return left;
    }
};